### PR TITLE
Always apply Python unixcompilers.py `rpath` fix, not just on cray

### PR DESCRIPTION
**Pull request by climbfuji**
## Description

Based on the conversation in the spack slack channel (https://spackpm.slack.com/archives/C5W7NKZJT/p1694393880262699), it was concluded that it would be safe to unconditionally apply the patch for the `rpath` linker flag in spack's Python installation, not just when the system is detected as Cray. The need for this came up when building on a Cray that was detected as a "new" Cray due to its `craype` version, but that still required using the Cray compiler wrappers.

Since there are no compilers in spack that understand `-R` but don't understand `-Wl,-R`, it's safe to apply this patch regardless of the platform.

Actual change: rename `var/spack/repos/builtin/packages/python/cray-rpath-3.1.patch` as `var/spack/repos/builtin/packages/python/rpath-non-gcc.patch` and apply unconditionally.

## Testing

Tested on the Navy's Narwhal machine (Cray) and a couple of other HPCs (non Cray) with GNU and Intel.

### CODE REVIEW

File: var/spack/repos/builtin/packages/python/package.py [hunk 1]
```diff
@@ -286,8 +286,8 @@ class Python(Package):
     patch("tkinter-3.10.patch", when="@3.10.0:3.10 ~tkinter")
     patch("tkinter-3.11.patch", when="@3.11.0:3.11 ~tkinter")
 
-    # Ensure that distutils chooses correct compiler option for RPATH on cray:
-    patch("cray-rpath-3.1.patch", when="@3 platform=cray")
+    # Ensure that distutils chooses correct compiler option for RPATH:
+    patch("rpath-non-gcc.patch")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/python/package.py]:
```suggestion
    patch("rpath-non-gcc.patch", when="@:3.11")
```
The distutils module will be removed in Python 3.12 so there won't be anything to patch.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Review by adamjstewart — APPROVED** [review]

**Comment by alecbcs** [comment]
@spackbot run pipeline

**Comment by scheibelp** [comment]
@spackbot run pipeline

**Comment by scheibelp** [comment]
@spackbot run pipeline

**Comment by scheibelp** [comment]
@spackbot run pipeline

**Pull request merged**